Docs caching (#7553)
solo-io/gloo: docs/content/guides/traffic_management/listener_configuration/caching/about.md

---
title: About caching responses
description: Learn about how caching works in Gloo Edge Enterprise with and without response validation. 
weight: 10
---

> With response caching, you can significantly reduce the number of requests Gloo Edge must make to an upstream service to return a response to a client.

Reviewer: Super Nit: Could we change this to something like 

> With response caching, you can significantly reduce the number of requests Gloo Edge makes to its upstream services

or

> By configuring response caching, Gloo Edge can significantly reduce the load on your upstream services.

> The Gloo Edge Enterprise caching filter is an extension (implementing filter) of the [Envoy cache filter](https://www.envoyproxy.io/docs/envoy/latest/start/sandboxes/cache) and takes advantage of all the cache-ability checks that are applied. However, Gloo Edge also provides the ability to store the cached objects in a Redis instance, including Redis configuration options such as setting a password.

Reviewer: Could we consider something like:
> The Gloo Edge Enterprise caching filter is an extension built on top of the [Envoy cache filter](https://www.envoyproxy.io/docs/envoy/latest/start/sandboxes/cache), and it is capable of all the functionality that the Envoy cache filter exposes. In addition, Gloo Edge provides the ability to store the cached objects in a Redis instance, including Redis configuration options such as setting a password.

I don't love my suggestion and I'm struggling with revising this paragraph, but I think it is not quite clear right now

> Use the following links to learn more about how caching works in Gloo Edge: 

Author: I removed the links

- [Caching without response validation](#caching-unvalidated)
- [Caching with response validation](#caching-validated)

## Caching without response validation {#caching-unvalidated}

The following diagram shows how response caching works without validation. 

![Caching without validation]({{% versioned_link_path fromRoot="/img/caching-unvalidated.svg/" %}})

1. The gateway forwards incoming requests to the upstream service where the request is processed. When the upstream service sends back a response to the client, the response is cached by the caching server. 

> 2. Subsequent requests from clients are not forwarded to the upstream. Instead, clients receive the cached response with an `age` response header from the caching server directly. The `age` response header shows the number of seconds that passed since the original response was sent. By default, responses are cached for 1 hour, unless the client specified a different time by sending the `cache-control` request header. After the time has passed and the cached response becomes stale, requests are forwarded to the upstream service again and a new response is cached by the caching server. 

Reviewer 2: ```suggestion
2. Subsequent requests from clients are not forwarded to the upstream. Instead, clients receive the cached response with an `age` response header from the caching server directly. The `age` response header shows the number of seconds that passed since the original response was sent. By default, responses are cached for 1 hour, unless the client specified a different time by sending the `cache-control` request header. After the time passes and the cached response becomes stale, requests are forwarded to the upstream service again and a new response is cached by the caching server. 
```

> 2. Subsequent requests from clients are not forwarded to the upstream. Instead, clients receive the cached response with an `age` response header from the caching server directly. The `age` response header shows the number of seconds that passed since the original response was sent. By default, responses are cached for 1 hour, unless the client specified a different time by sending the `cache-control` request header. After the time has passed and the cached response becomes stale, requests are forwarded to the upstream service again and a new response is cached by the caching server. 

Reviewer: > By default, responses are cached for 1 hour, unless the client specified a different time by sending the `cache-control` request header.

I think this is not quite accurate -- I think responses are only cached if the upstream service returns a `cache-control` header, and then the responses are cached for the amount of time specified by that header. Could we instead say something like

> 1. The gateway forwards incoming requests to the upstream service where the request is processed. When the upstream service sends a response to the client, it can indicate that the response is cacheable by setting the `cache-control` response header. When Gloo Edge receives a response with a `cache-control` header, that response is cached by the caching server for the amount of time specified in the header.
> 2. Subsequent requests from clients are not forwarded to the upstream. Instead, clients receive the cached response with an `age` response header from the caching server directly. The `age` response header shows the number of seconds that passed since the original response was sent. Responses are cached for the amount of time specified in the `cache-control` response header. After the time has passed and the cached response becomes stale, requests are forwarded to the upstream service again and a new response is cached by the caching server.

## Caching with response validation {#caching-validated}

The following diagram shows how response caching works when the upstream service supports response validation. 

![Caching with validation]({{% versioned_link_path fromRoot="/img/caching-validated.svg/" %}})

1. The gateway forwards incoming requests to the upstream service where the request is processed. When the upstream service sends back a response to the client, the response is cached by the caching server. 

> 2. Subsequent requests from clients are not forwarded to the upstream. Instead, clients receive the cached response with an `age` response header from the caching server directly. The `age` response header shows the number of seconds that passed since the original response was sent. By default, responses are cached for 1 hour, unless the client specified a different time by sending the `cache-control` request header. 

Reviewer: Similarly to above, the information about the default cache time of 1 hour is not correct. Could we do:
> 1. The gateway forwards incoming requests to the upstream service where the request is processed. When the upstream service sends a response to the client, it can indicate that the response is cacheable by setting the `cache-control` response header. When Gloo Edge receives a response with a `cache-control` header, that response is cached by the caching server for the amount of time specified in the header.
> 2. Subsequent requests from clients are not forwarded to the upstream. Instead, clients receive the cached response with an `age` response header from the caching server directly. The `age` response header shows the number of seconds that passed since the original response was sent. Responses are cached for the amount of time specified in the `cache-control` response header.

> 3. After the time has passed and the cached response becomes stale, the response validation period starts. In order for response validation to work, upstream services must be capable of processing `If-Modified-Since` request headers that are sent from the client. If the upstream's response changed since the time that is specified in the `If-Modified-Since` request header, the new response is forwarded to the client and cached by the caching server (3a). Subsequent requests receive the cached response until the cache timeframe has passed again (2). If the response has not changed, the upstream service sends back a 304 Not Modified HTTP respones code. The gateway then gets the cached response from the caching server and returns it to the client (3b). Response validation continues for subsequent requests until a new response is received from the upstream service. 

Reviewer 2: typo, I wasn't completely sure what (2) meant, and maybe also make sub-steps to match the diagram
```suggestion
3. After the time passes and the cached response becomes stale, the response validation period starts. In order for response validation to work, upstream services must be capable of processing `If-Modified-Since` request headers that are sent from the client. 
   1. If the upstream's response changed since the time that is specified in the `If-Modified-Since` request header, the new response is forwarded to the client and cached by the caching server (3a). Subsequent requests receive the cached response until the cache timeframe passes again (as shown in Step 2). 
   2. If the response did not change, the upstream service sends back a `304 Not Modified` HTTP response code. The gateway then gets the cached response from the caching server and returns it to the client (3b). Response validation continues for subsequent requests until a new response is received from the upstream service. 
```

---
title: About caching responses
description: Learn about how caching works in Gloo Edge Enterprise with and without response validation. 
weight: 10
---

With response caching, you can significantly reduce the number of requests Gloo Edge makes to its upstream services.

{{% notice note %}}
This feature is available only for Gloo Edge Enterprise v1.12.x and later.
{{% /notice %}}

> The Gloo Edge Enterprise caching filter is an extension built on top of the Envoy cache filter, and it is capable of all the functionality that the Envoy cache filter exposes. In addition, Gloo Edge provides the ability to store the cached objects in a Redis instance, including Redis configuration options such as setting a password.

Reviewer: This paragraph is duplicated on lines 13 and 15 -- can we remove one of them?

The Gloo Edge Enterprise caching filter is an extension that is built on top of the [Envoy cache filter](https://www.envoyproxy.io/docs/envoy/latest/start/sandboxes/cache), and includes all of the functionality that the Envoy cache filter exposes. In addition, Gloo Edge provides the ability to store the cached objects in a Redis instance, including Redis configuration options such as setting a password.

Review the information on this page to learn more about how caching works in Gloo Edge. 

## Caching without response validation {#caching-unvalidated}

The following diagram shows how response caching works without validation. 

![Caching without validation]({{% versioned_link_path fromRoot="/img/caching-unvalidated.svg/" %}})

> 1. When the gateway receives an incoming request from a client, it checks with the caching server if a cached response for the referenced upstream is available. Because no cached response is available, the request is forwarded to the upstream service where the request is processed. When the upstream service sends a response to the client, the upstream can indicate if the response is cacheable by providing the `cache-control` response header. When Gloo Edge receives a response with a `cache-control` header, the response is cached by the caching server for the amount of time that is specified in the header. 

Reviewer: ```suggestion
1. When the gateway receives an incoming request from a client, it checks with the caching server if a cached response for the referenced upstream is available. Because no cached response is initially available, the request is forwarded to the upstream service where the request is processed. When the upstream service sends a response to the client, it can indicate if the response is cacheable by providing the `cache-control` response header. When Gloo Edge receives a response with a `cache-control` header, the response is cached by the caching server for the amount of time that is specified in the header. 
```

5. Subsequent requests from clients are not forwarded to the upstream. Instead, clients receive the cached response with an `age` response header from the caching server directly. The `age` response header shows the number of seconds that passed since the original response was sent. After the time that was specified in the `cache-control` header passes, the cached response becomes stale. Requests are then forwarded to the upstream service again and a new response is sent to the client.

## Caching with response validation {#caching-validated}

The following diagram shows how response caching works when the upstream service supports response validation. 

![Caching with validation]({{% versioned_link_path fromRoot="/img/caching-validated.svg/" %}})

> 1. When the gateway receives an incoming request from a client, it checks with the caching server if a cached response for the referenced upstream is available. Because no cached response is available, the request is forwarded to the upstream service where the request is processed. When the upstream service sends a response to the client, the upstream can indicate if the response is cacheable by providing the `cache-control` response header. When Gloo Edge receives a response with a `cache-control` header, the response is cached by the caching server for the amount of time that is specified in the header. 

Reviewer: ```suggestion
1. When the gateway receives an incoming request from a client, it checks with the caching server if a cached response for the referenced upstream is available. Because no cached response is initially available, the request is forwarded to the upstream service where the request is processed. When the upstream service sends a response to the client, the upstream can indicate that the response is cacheable by providing the `cache-control` response header. When Gloo Edge receives a response with a `cache-control` header, the response is cached by the caching server for the amount of time that is specified in the header. 
```
